sdk/nodejs/runtime/invoke.ts

```diff
@@ -251,28 +251,35 @@ export function call<T>(tok: string, props: Inputs, res?: Resource): Output<T> {
 }
 
 /**
- * Behaves exactly like `call` but returns a Promise instead. Unknowns are not allowed in the response, and an exception
- * will be thrown blaming the provider for violating the contract if the provider returns any unknown values. Secret
- * bits and dependencies returned from the provider are similarly discarded.
+ * Behaves exactly like `call` but returns a Promise instead, to support code generated calls for methods with
+ * XReturnPlainResource. After awaiting the output, extract the resource by field name specified in plainResourceField.
+ * Unknowns are not allowed in the response, and an exception will be thrown blaming the provider for violating the
+ * contract if the provider returns any unknown values. Secret bits and dependencies returned from the provider are
+ * similarly discarded at the moment.
  */
-export function callAsync<T>(tok: string, props: Inputs, res?: Resource): Promise<T> {
+export function callAsync<T>(tok: string, props: Inputs, res: Resource, callAsyncOpts: {
+    plainResourceField: string,
+}): Promise<T> {
     const label = `Calling function: tok=${tok} (callAsync)`;
     log.debug(label + (excessiveDebugOutput ? `, props=${JSON.stringify(props)}` : ``));
 
     return debuggablePromise(new Promise<T>((resolve, reject) => {
-        const resolver = (v: T, _isKnown: boolean, _isSecret: boolean, _deps?: Resource[], err?: Error | undefined) => {
+        const resolver = (v: T, isKnown: boolean, _isSecret: boolean, _deps?: Resource[], err?: Error | undefined) => {
             if (err) {
                 return reject(err);
             }
 
-            // For objects we recur into each field and transform it into an Output, which recursively flattens any
-            // nested Output. This is necessary for type safety, since the expected return types from Call methods
-            // expose properties as prompt values, but the runtime values such as v.foo here are Outputs.
-            //
-            // Unlike call, callAsync does not just transform v into an Output at top level to avoid unkown poisoning,
-            // the situation when any output field being unknown results in the entire result being unknown and
-            // corresponding loss of information. Instead, unknown values are discarded.
-            unsafeUnwrapProperties(v).then(resolve);
+            const extracted = (<any>v)[callAsyncOpts.plainResourceField];
+
+            if (isUnknown(extracted)) {
```

Comment:
@Frassle so this is the interesting case. I can hit this case at runtime when I modify the meta-provider to hard-return an UNKNOWN instead of returning an AWS Provider. 

Like this:

```go
func CallConfigureAwsMethod(ctx *pulumi.Context, inputs provider.CallArgs) (*provider.CallResult, error) {
	if ctx.DryRun() {
		// Really return unknown.
		result := &ConfigureAwsMethodResult{
			AwsProvider: pulumi.UnsafeUnknownOutput(nil).ApplyT(func(x any) *aws.Provider {
				panic("This should not be called")
			}).(aws.ProviderOutput),
		}
		return provider.NewCallResult(result)
	}
```

Note in this case the meta-provider didn't even allocate the AWS provider so there's no URN in the system.

IN this case, we must have an error here. If we do not inject an error, we get an obscure error down the line.

```
    TypeError: provider.getPackage is not a function
        at convertToProvidersMap (/Users/anton.tayanovskyy/code/pulumi-12709/pulumi/sdk/nodejs/resource.ts:517:25)
        at new Resource (/Users/anton.tayanovskyy/code/pulumi-12709/pulumi/sdk/nodejs/resource.ts:439:16)
        at new CustomResource (/Users/anton.tayanovskyy/code/pulumi-12709/pulumi/sdk/nodejs/resource.ts:880:9)
        at new Bucket (/Users/anton.tayanovskyy/code/pulumi-12709/examples/ts-example/node_modules/@pulumi/s3/bucket.ts:522:9)
        at file:///Users/anton.tayanovskyy/code/pulumi-12709/examples/ts-example/index.ts:22:16
        at processTicksAndRejections (node:internal/process/task_queues:95:5)
```

If we want to detect and error out, there's two approaches. One is to detect in every language SDK like this code is written. Another is to perhaps modify github.com/pulumi/pulumi/pkg/v3/resource/provider package to be more schema aware and blame the user code in the Go layer for violating an invariant.. That sounds like a potentially better way to enforce, one place to edit instead of per language.

I do think it should be an error for something that wants to return a "plain" resource to return an unknown without even a URN - this is very different from returning a resource with unknown outputs. No recovery is possible. Having the signature return Promise<T> means this must be an error if it happens.